### PR TITLE
[master] stalekey engine fixes

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?

Fixing stalekey engine so it deletes the keys when they are a list.

Breaking out various functionality so it's easier to test.

Writing tests.

### What issues does this PR fix or reference?
#55057 

### Tests written?
**[NOTICE] Bug fixes or features added to Salt require tests.**
Please review the [test documentation](https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite.
Yes
### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/engines/stalekey.py [hunk 1]
```diff
@@ -39,41 +39,103 @@
...
+    '''
+    Indicate if engine should continue running
+    Useful when running tests and we want to ensure
+    the engine only makes one pass
+    '''
+    return True
+
+
 def start(interval=3600, expire=604800):
+    '''
```
Comment (s0undt3ch) [inline salt/engines/stalekey.py]: I'm going to pull a @dwoz on you :)
For reference, https://github.com/saltstack/salt/pull/55047#discussion_r336696340

Let's make start accept `_check_running_callable=None`

Then the `while` code Bellow becomes:
```python
while True:
    # do stuff
    If callable(_check_running_callable) and _check_running_callable() is False:
        break
```

No?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]
You need to pull master and rebase too.

File: salt/engines/stalekey.py (hunk 1 above)
Comment (garethgreenaway) [reply salt/engines/stalekey.py]: Not against the idea, just curious, what is the advantage?

Comment (garethgreenaway) [reply salt/engines/stalekey.py]: Actually the `_running` function was being used when I was attempting to run `start` from tests, which didn't work out.  So I can remove that.

Comment (s0undt3ch) [reply salt/engines/stalekey.py]: Not having a function for us just to patch during tests, ie, dependency injection.

**Review by s0undt3ch — APPROVED** [review]

**Review by dwoz — APPROVED** [review]

File: tests/unit/engines/test_stalekey.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+import logging
+import msgpack
+
+# Import Salt Testing Libs
+from tests.support.mixins import LoaderModuleMockMixin
+from tests.support.unit import skipIf, TestCase
+from tests.support.mock import (
+    NO_MOCK,
+    NO_MOCK_REASON,
```
Comment (waynew) [inline tests/unit/engines/test_stalekey.py]: Gonna need to remove `NO_MOCK` and `NO_MOCK_REASON` here.

File: tests/unit/engines/test_stalekey.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+class MockKey(object):
+    def __init__(self, opts):
+        pass
+
+    def delete_key(self, key):
+        pass
+
+
+@skipIf(NO_MOCK, NO_MOCK_REASON)
```
Comment (waynew) [inline tests/unit/engines/test_stalekey.py]: This no mock skip should also be removed

**Pull request closed without merging**